Adding a basic daily-js demo

markbackman opened the pull request:
I'm adding a vanilla JS demo showing how to get started with daily-js. We have tons of more sophisticated examples. This one is meant to be as simple as possible. I plan to add a getting started guide to the Guides section in the docs.

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.html:
```diff
@@ -0,0 +1,51 @@
+        id="room-url"
+        size="50"
+        placeholder="https://yourcompany.daily.co/hello"
+        value="https://mark.staging.daily.co/waiting_1"
```
I think I'd suggest not filling it in with a valid default room URL, unless you _clearly_ call out that the user trying this demo may land in a room with a stranger in it.

markbackman replied:
Thanks for catching this! 😅

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+   * @param {Object} event - The track event object containing the video track.
+   */
+  displayVideo(event) {
+    let videosDiv = document.getElementById('videos');
```
```suggestion
    const videosDiv = document.getElementById('videos');
```

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+   */
+  displayVideo(event) {
+    let videosDiv = document.getElementById('videos');
+    let videoEl = document.createElement('video');
```
```suggestion
    const videoEl = document.createElement('video');
```

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+    if (event.participant.local) {
+      return; // Avoid playing local audio
+    }
+    let audioEl = document.createElement('audio'); // Create <audio> element
```
```suggestion
    const audioEl = document.createElement('audio'); // Create <audio> element
```

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+   * Adds a <video> element to the DOM and attaches the video track to it.
+   * @param {Object} event - The track event object containing the video track.
+   */
+  displayVideo(event) {
```
This looks to create a new video div each time a track starts - what happens if the track changes when a track was already started before, issuing a new `"track-started"` event for the same participant? (e.g., in case of device change or meeting move)

markbackman replied:
The `videos` div is constant. I'm creating a new `<video>` element which I append to the `videos` div. In testing it, it does what I would expect. I think this is OK, but let me know if you have any concerns.

lazeratops replied:
Ah sorry I misspoke when I said the div. I did mean the video element. Basically, what happens if a participant who already _has_ a video element here in this div gets _another_ `"track-started"` event for any reason? (I'm not sure if you permit device switching here but specifically a meeting move comes to mind, which has bitten me in demos when I don't account for it before!). It seems like we will end up appending _another_ video element for the same participant without destroying the old one, no?

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+   */
+  destroyTrack(event) {
+    let tracks = document.querySelectorAll('video, audio');
+    tracks.forEach((el) => {
```
I think we need to be sure to set `srcObject` to `null` to avoid leaving `WebMediaPlayer`s dangling indefinitely. E.g.: https://github.com/daily-demos/modern-wordfare/pull/22

>

markbackman replied:
Good catch. Thank you 🙏

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+    const joinOptions = { url: roomUrl };
+    if (joinToken) {
+      joinOptions.token = joinToken;
+      console.log('Joining with token:', joinToken);
```
Suggest not printing access token to console

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+
+    try {
+      await this.call.join(joinOptions);
+      console.log(`Successfully joined: ${roomUrl}`);
```
Just a thought: I wonder if using the `"joined-meeting"` event instead of printing this after the await might be nicer and more consistent. Then we show _one_ consistent flow for when a participant joins a meeting, instead of splitting it between the event listener _and_ await approaches.

markbackman replied:
Good call 👍

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+   */
+  getTrackStates() {
+    if (this.call && this.call.participants().local) {
+      const localParticipant = this.call.participants().local;
```
Why don't we define this up above the if statement, so we don't invoke `this.call.participants().local` twice?

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+  getTrackStates() {
+    if (this.call && this.call.participants().local) {
+      const localParticipant = this.call.participants().local;
+      this.isCameraMuted = localParticipant.tracks.video.state === 'off';
```
For readability, might be nice to define a `tracks` const for reuse in these two lines: `const tracks = localParticipant.tracks`

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+   */
+  async leave() {
+    try {
+      await this.call.leave();
```
Wondering the same thing here  as the join handling - maybe worth using the `"left-meeting"` event here?

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,251 @@
+
+      // Remove video and audio elements
+      document.querySelectorAll('#videos video, audio').forEach((el) => {
+        el.remove();
```
Need to remember setting source objects to null here as well

lazeratops reviewed (commented):
Pre-emptive warning that I have _not run this yet_, only read through the code! Looks great, left some small suggestions.

markbackman commented:
@lazeratops this should be ready for review again. I addressed your feedback with the exception of the one we're still discussing. I also opted to add a device selector so that I can call the related methods specifically in the guide.

Thanks again for the review! 🙌

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.html:
```diff
@@ -0,0 +1,59 @@
+        type="text"
+        id="join-token"
+        size="50"
+        placeholder="Enter Join Token (optional)"
```
Sorry I didn't notice this before, but I wonder if we should actually make this say "Enter Daily meeting token" (and link the text to our meeting token docs or guide). To remain consistent and clear with the terminology.

markbackman replied:
Good idea!

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,389 @@
+   * Handler for remote participants joining and leaving. Updates the participant count.
+   */
+  handleParticipantJoinLeft() {
+    document.getElementById(
```
Optional nit: since you're duplicating this here and on line 70 might be nice to pull it out into a little method for reuse.

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,389 @@
+   * @param {Object} event - The track event object.
+   */
+  destroyTrack(event) {
+    let tracks = document.querySelectorAll('video, audio');
```
```suggestion
    const tracks = document.querySelectorAll('video, audio');
```

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -0,0 +1,389 @@
+    tracks.forEach((el) => {
+      if (el.srcObject && el.srcObject.getTracks().includes(event.track)) {
+        el.srcObject = null; // Release media resources
+        el.remove(); // Remove the element from the DOM
```
One thought after running this, but may be out of scope so I will defer to you! Currently if you are testing with more than one participant and turn the camera off it can get a little confusing to see the visual representation of the other participant _completely_ disappear. We DO have the participant counter which helps, but another thing that might help is keeping a div for the participant with a background color, or just their session ID, or something to indicate another presence "tile" that is just toggled off. Totally optional like I said, might just be a little more intuitive with how people are used to seeing camera-off visuals in video calls.

lazeratops reviewed (approved):
LGTM with just a couple of optional nits.

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -139,13 +158,49 @@ class DailyCallManager {
-    const videoEl = document.createElement('video');
-    videosDiv.appendChild(videoEl);
-    videoEl.style.width = '100%';
+    const participantId = event.participant.session_id; // Assume this is how we get the participant ID
```
Tempted to suggest just cutting this comment as for someone new to Daily and not yet thinking about _other_ ways to get the participant ID it might be a little confusing.

markbackman replied:
Whoops! That's my own comment left over. 🙈 Removing.

lazeratops commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -165,18 +220,47 @@ class DailyCallManager {
+    if (effectiveParticipantId) {
+      this.removeMediaElement(effectiveParticipantId, event.track.kind);
+    } else {
+      console.error(
```
Sort of related to my question about participants leaving, I am consistently getting this error when a participant leaves. This makes me think that the participant leave event is firing before the track stopped event, which I suspect also results in the associated WebMediaPlayer only being paused, not destroyed: 

<img width="827" alt="image" src="https://github.com/daily-demos/daily-samples-js/assets/65890040/fd36bc13-12c0-4884-8ac6-b25a02f61d61">

markbackman replied:
You're right. This isn't working upon leaving. The root cause appears to be a bug in daily-js, which I just discovered thanks to your nudging: https://linear.app/dailyco/issue/CSDK-2110/when-a-track-stopped-event-is-emitted-due-to-a-participant-leaving-a.

I'm inclined to leave this in here and just fixing this issue in daily-js. With the participant information missing from track-stopped, the simplicity of `track-stopped` is broken. This forces someone to rely on `participant-updated` to get the full event. Anyway, thanks for finding this issue! 🙌

lazeratops replied:
Makes sense! Approving

markbackman commented on samples/client-sdk/getting-started-daily-js/index.js:
```diff
@@ -67,18 +69,29 @@ class DailyCallManager {
+          `video-container-${participantId}`
+        );
+        if (videoContainer) {
+          videoContainer.remove(); // Removes the video container from the DOM.
```
:/
https://github.com/daily-demos/daily-samples-js/pull/17#discussion_r1471877843

lazeratops reviewed (approved):
🚢
